src/box/tuple_compare.cc

```diff
@@ -1429,6 +1433,36 @@ func_index_compare(struct tuple *tuple_a, hint_t tuple_a_hint,
 						cmp_def);
 	if (rc != 0)
 		return rc;
+	/*
+	 * If key_a and key_b are equal, the index is unique and there is a
+	 * nullable key part, then two options are possible:
+	 * 1. some part is NULL in both keys => compare primary keys;
+	 * 2. there are no NULL parts => the tuples are really equal.
+	 * Note that a part cannot be NULL in one key and non-NULL in another
+	 * when key_compare_parts() returns 0.
+	 *
+	 * The index is unique and nullable when cmp_def->part_count is greater
+	 * than cmp_def->unique_part_count. In all other cases they are equal.
+	 */
+	bool is_unique_nullable = cmp_def->part_count >
+				  cmp_def->unique_part_count;
+	if (is_unique_nullable) {
+		assert(is_nullable);
+		bool was_null_met = false;
+		for (uint32_t i = 0; i < key_part_count; i++) {
+			if (mp_typeof(*key_a) == MP_NIL) {
+				was_null_met = true;
```

Comment:
1. `key_a` and `key_b` are not used later, so we can break the loop after the first `nil`.
2. Nit: subject-predicate form is easier to comprehend (`was_null_met` -> `null_was_met`).
3. I don't like the 2nd pass to find out if there were any nils. However, it makes it easier that we can stop the pass after the first `nil`. This code is simple and straightforward, which is good, but adding a new output argument to `key_compare_and_skip_parts` would be a more effective solution. I would ask @alyapunov

Reply:
1. Done;
2. I want to keep the name as in `tuple_compare_slowpath()` for easier grepping.

Reply:
What's about the third point?

Reply:
Discussed f2f, we will leave it up to the second reviewer.

Reply:
Up to @locker

Reply:
Agree with @drewdzzz. The code would be more straightforward if we inlined key comparison in this function and set the `was_null_met` flag in the loop. We wouldn't even need this verbose comment then.

Reply:
Done. The comment is still relevant though.